internal/context/options.go

```diff
@@ -0,0 +1,33 @@
+package context
+
+// Option allows configuring the context
+type Option func(*context)
+
+// WithMethod sets the Context method name
+func WithMethod(method string) Option {
+	return func(ctx *context) {
+		ctx.method = method
+	}
+}
+
+// WithArguments sets the Context arguments
+func WithArguments(args map[string]interface{}) Option {
+	// TODO this should loop and add all fields instead of overwriting the map
+	return func(ctx *context) {
+		ctx.arguments = args
+	}
+}
+
+// WithCorrelationID sets the CorrelationID
+func WithCorrelationID(cID string) Option {
+	return func(ctx *context) {
+		ctx.correlationID = cID
+	}
+}
+
+// WithArgument sets a Context argument
+func WithArgument(key string, value interface{}) Option {
+	return func(ctx *context) {
+		ctx.arguments[key] = value
```

Comment:
Again, I might be wrong here, but to me looks like the way we write and read arguments, is not goroutine safe, as I believe a mutex lock is needed to ensure that when reading/writing to a map.

If I'm right, doesn't this go against the core idea of context's being goroutine safe?

Reply:
context is immutable, this way we never update the context after it has been created.
`WithArguments` is an option that is used on `.New()` given an existing context and returns a new context. It does __not__ modify the old context.

Reply:
Ah, that makes sense, at least for `WithArguments()`, but `WithArgument()` here does modify the map. But I assume it's only meant to be used together with `New()` right?

Isn't this a bit different than the various `With*` method on the stdlib context package? They all return a new context if I've understood it correctly. That does however allocate a new context object for each new value you add, so is that why the methods here effectively return `Option` callbacks which are all passed and applied to the new one context instance via `New()`?

Reply:
The `New()` is basically accepting a number functional options to allow multiple things to be set at the same context without having to create  a new one each time. `https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis`, `https://medium.com/soon-london/variadic-configuration-functions-in-go-8cef1c97ce99`.

The methods basically return a function that is called from within `New()` and they are given the new context created. In both cases "modifying the map" only happens within the scope of `New()` and cannot happen concurrently with anything else.